content/blog/2020-12-18-futil-passes/index.md

```diff
@@ -0,0 +1,146 @@
+
+For some `cell` definitions, we know statically the number of cycles it takes from the moment a `go` signal is raised to when the `done` signal is raised. For registers, this latency is a single cycle. Starting with this knowledge, it's possible to deduce the latencies of some `group`s. This pass targets `group`s where these conditions are true:
+- All `go` ports are written to only by `done` ports or constants.
+- All `done` ports are only written to by other `done` signals.
+- The latency of each cell in this chain of `go`/`done` signals is known statically.
+
+For example, this group, called `a`, takes only a single cycle to execute:
+
+    group a {
+      register1.write_en = 1'd1;
+      a[done] = register1.done;
+    }
+
+The `go` port of `register1` is driven by a constant, and the `done` signal of the group is written by the `done` signal of `register1`.
+
+This idea can be extended to form these sort of `go`/`done` chains:
+
+    group b {
+      r1.write_en = 1'd1;
+      r2.write_en = r1.done;
+      r3.write_en = r2.done;
+      b[done] = r3.done;
+    }
+
+The pass implementation infers latencies for these groups with a graph-based approach. The FuTIL codebase provides a convenient library for generating graphs representing different parts of FuTIL programs. The inference pass uses this graph library to generate a graph with the following algorithm:
+
+1. Create nodes for every port in the group.
+2. Draw an edge between each node $(u, v)$ if $u$ writes to $v$.
+3. Draw an edge between each node $(u, v)$ if $u$ is a `done` signal, and $v$ is a `go` signal.
+4. Delete isolated vertices (vertices with indegree and outdegree $0$).
+5. Weight each edge after a `go` port with the latency of its parent `cell`.
+
+For example, the group `b` would have this graph built from it:
+
+<img src="graph.png" style="width:650px;height:350px"/>
+
+The pass then simply traverses the path between the start node and end node (which can be found by computing the first and last elements of a topological sort), summing the edge weights along the way. If there is no path, then this sort of `go`/`done` chain pattern isn't in the `group`, and the pass conservatively gives up.
+
+### Evaluation
+
+I added some simple unit tests to the FuTIL test suite, and also ran the pass on programs generated by a systolic array-to-FuTIL compiler, which does not insert static timing annotations. The generated programs continued to pass correctness tests, providing some confidence that the inference pass implementation is correct.
```

Comment:
Maybe report the total number of programs you tried?

Reply:
Reported on the specific number of programs tried.